src/remote_interface.c

```diff
@@ -0,0 +1,262 @@
+    memcpy(messagebody + TRANSACTION_TRYTES_LENGTH, buf, 4);
+
+
+    if (!declare_callback_queue(&remoteinterface_ctx->conn, channel,
+                                &reply_to_queue))
+        goto fail;
+
+    if (!publish_message_with_relpy_to(&remoteinterface_ctx->conn, channel,
+                                       "incoming_queue", reply_to_queue,
+                                       messagebody))
+        goto fail;
+
+    if (!wait_response_message(&remoteinterface_ctx->conn, channel,
+                               reply_to_queue, (char *) (ctx->output_trytes),
+                               TRANSACTION_TRYTES_LENGTH))
+        goto fail;
+
+    amqp_bytes_free(reply_to_queue);
+
+    PoWValidation(ctx->output_trytes, ctx->mwm);
+
+    return true;
+
+fail:
+    return false;
+}
+
+static bool PoWRemoteInterface_Context_Initialize(
+    RemoteInterfaceContext *remoteinterface_ctx)
+{
+    if (remoteinterface_ctx->num_max_thread <= 0)
+        goto fail_to_init;
+
+    PoW_Remote_Interface_Context *ctx = (PoW_Remote_Interface_Context *) malloc(
+        sizeof(PoW_Remote_Interface_Context) *
+        remoteinterface_ctx->num_max_thread);
+
+    memset(remoteinterface_ctx->slots, 0,
+           remoteinterface_ctx->num_max_thread * sizeof(bool));
+
+    if (!connect_broker(&remoteinterface_ctx->conn,
+                        remoteinterface_ctx->num_max_thread))
+        goto fail_to_init;
+
+    if (!declare_queue(&remoteinterface_ctx->conn, 1, "incoming_queue"))
+        goto fail_to_init;
+
+    remoteinterface_ctx->context = ctx;
+
+    uv_mutex_init(&remoteinterface_ctx->lock);
+
+    return true;
+
+fail_to_init:
+    return false;
+}
+
+static void PoWRemoteInterface_Context_Destroy(
+    RemoteInterfaceContext *remoteinterface_ctx)
+{
+    PoW_Remote_Interface_Context *ctx =
+        (PoW_Remote_Interface_Context *) remoteinterface_ctx->context;
+
+    disconnect_broker(&remoteinterface_ctx->conn,
+                      remoteinterface_ctx->num_max_thread);
+
+    free(ctx);
+}
+
+static void *PoWRemoteInterface_getPoWContext(
+    RemoteInterfaceContext *remoteinterface_ctx,
+    int8_t *trytes,
+    int mwm)
+{
+    uv_mutex_lock(&remoteinterface_ctx->lock);
+
+    for (int i = 0; i < remoteinterface_ctx->num_max_thread; i++) {
+        if (!remoteinterface_ctx->slots[i]) {
+            remoteinterface_ctx->slots[i] = true;
+
+            uv_mutex_unlock(&remoteinterface_ctx->lock);
+            PoW_Remote_Interface_Context *ctx =
+                remoteinterface_ctx->context +
+                sizeof(PoW_Remote_Interface_Context) * i;
+            memcpy(ctx->input_trytes, trytes, TRANSACTION_TRYTES_LENGTH);
+            ctx->mwm = mwm;
+            ctx->indexOfContext = i;
+
+            return ctx;
+        }
+    }
+
+    uv_mutex_unlock(&remoteinterface_ctx->lock);
+
+    return NULL; /* It should not happen */
+}
+
+static bool PoWRemoteInterface_freePoWContext(
+    RemoteInterfaceContext *remoteinterface_ctx,
```

Comment:
`remoteinterface_ctx` might not a good name. Instead, simply using `ctx` should work out of box.

Reply:
In the remote_interface.c, `ctx` name has been used.  Therefore, for simplicity, `remoteinterface_ctx ` will be renamed into `remote_ctx`.

Reply:
By the way, the naming `PoWRemoteInterface_freePoWContext` is really redundant. It was intended to be a helper function . Simply calling it as `RI_freePoWcontext` works well.